core/arch/arm/tee/drv_mgt.c

```diff
@@ -0,0 +1,344 @@
+#endif
+	return TEE_SUCCESS;
+}
+
+/**
+ * @brief   Installation of all registered drivers.\n
+ *          Add all drivers in the drivers list in order
+ *          to maintain drivers function of system power.
+ *
+ * @retval TEE_SUCCESS               Success
+ * @retval TEE_ERROR_OUT_OF_MEMORY   Out of memory
+ */
+static TEE_Result do_install(void)
+{
+	struct drv_ref *new_elem;
+	const struct driver *drv = NULL;
+
+	const struct driver *start = SCATTERED_ARRAY_BEGIN(
+						drivers, struct driver);
+	const struct driver *end = SCATTERED_ARRAY_END(
+						drivers, struct driver);
+
+	for (drv = start; drv < end; drv++) {
+		if (drv->ops) {
+			if (!drv->ops->init) {
+				MGTDRV_TRACE(
+				"Driver [%s] not install no Init function",
+				drv->name);
+				continue;
+			}
+
+			MGTDRV_TRACE("Add driver [%s] in the list", drv->name);
+
+			/*
+			 * Allocate a new driver reference element to
+			 * be added in the driver list
+			 */
+			new_elem = malloc(sizeof(struct drv_ref));
+			if (!new_elem)
+				return TEE_ERROR_OUT_OF_MEMORY;
+
+			/*
+			 * Fill and add the new driver reference into the
+			 * list at the head of the list.
+			 */
+			new_elem->drv  = drv;
+			new_elem->data = NULL;
+			new_elem->mode = OFF;
+
+			if (drv->data_size) {
+				new_elem->data = drivers_alloc_data(
+							drv->data_size);
+				if (!new_elem->data)
+					return TEE_ERROR_OUT_OF_MEMORY;
+			}
+
+			SLIST_INSERT_HEAD(&drv_list, new_elem, next);
+		}
+#ifdef DRV_MGT_DEBUG
+		else
+			MGTDRV_TRACE("Driver [%s] OPS not defined", drv->name);
+#endif
+
+	}
+
+	IMSG("Drivers installation success");
+
+	return TEE_SUCCESS;
+}
+
+/**
+ * @brief   Initialization of all driver in the list.\n
+ *
+ * @note    Return TEE_SUCCESS in case there is no driver
+ *
+ * @retval TEE_SUCCESS       Success
+ * @retval TEE_ERROR_GENERIC Generic error
+ */
+static TEE_Result do_init(void)
+{
+	TEE_Result ret;
+	uint8_t error_counter = 0;
+
+	struct drv_ref *drv_elem;
+
+	SLIST_FOREACH(drv_elem, &drv_list, next) {
+		ret = drv_elem->drv->ops->init(drv_elem->data);
+		MGTDRV_TRACE("Driver [%s] init ret 0x%"PRIx32"",
+				drv_elem->drv->name, ret);
+		if (ret == TEE_SUCCESS)
+			drv_elem->mode = ON;
+		else {
+			SLIST_REMOVE(&drv_list, drv_elem, drv_ref, next);
+			free(drv_elem);
+			error_counter++;
+		}
+	}
+
+	return (error_counter) ? TEE_ERROR_GENERIC : TEE_SUCCESS;
+}
+
+/**
+ * @brief   Prepare all drivers to enter in a the given power \a mode.
+ *          Function of the \a wait parameter:
+ *           - \a wait = TRUE, the driver must complete the power mode
+ *           entry procedure before returning.
+ *           - \a wait = FALSE, the driver checks if it's ready or not.
+ *           If not, it return TEE_ERROR_BUSY, otherwise it switches in
+ *           the requested power mode.
+ *
+ * @note    Return TEE_SUCCESS in case there is no driver
+ *
+ * @param[in]  mode   Power mode to switch to
+ * @param[in]  wait   TRUE to wait until power mode reach
```

Comment:
i don't catch to goal of this argument and how `TEE_ERROR_BUSY` error is used.
Would you like the pm callback to be called again when a driver is busy?

Reply:
On complex system using HW DMA operation instantiated in the OPTEE, the transition to a power state could be:
 - Canceled because developper know that waiting DMA transaction completion is too long and there is interrest to do the power state transition
 - Or if this is a mandatory operation, the DMA transaction should be completed even if it takes time.

DMA is the typical usecase because OPTEE is never the initiator of the operation, this is Host OS that ask OPTEE to execute the DMA transaction (implicitly or not). Problem is that the Host OS doesn't know if OPTEE is using DMA and/or if the DMA is still running, Hence, if Host OS decides to switch off external memory/peripheral on which the DMA request is connected to, DMA will stall or even system can crash.
This is why in case of OPTEE-OS any power state transition must be approved by OPTEE.

Reply:
A driver could simply return from its pm_suspend callback with a error code that make the suspend sequence to fail (and restore the platform the running state) when it cannot suspend.

Reply:
Well, when we want to support the same driver on several device (because this is the same peripheral), It's easier to indicate to the driver what system is expecting. In that way driver can be generic and support any state.